### PR TITLE
UCP/CORE/INFO: Add support querying CM's UCP EP info

### PR DESCRIPTION
**Pull request by dmitrygx**
## What

Add support querying CM's UCP EP info/

## Why ?

`ucx_info` utility is useful to check the EP configuration on a concrete system, but lack of CM support doesn't allow to check CM UCP EP configuration.

## How ?

1. Update `ucx_info` utility to accept `-A <ip_addr>` which should be used to specify IP address to connect to.
2. Add support for CM flow: creating listener, connection using socket address to the listener.
3. Add UCP EP flush to make sure that UCP EP has a final configuration.
4. Update UCP EP control extension to save CM index to have the access to the value any time, i.e. even after CM WIREUP EP is not available anymore.
5. Remove CM index field from WIREUP EP - it is needed anymore, UCP EP control extension one will be used to determine CM index.
6. Update `ucp_ep_config_print()` to accept UCP EP instead of UCP configuration.
7. Update `ucp_ep_config_print()` to print CM lane information.

### CODE REVIEW

**Comment by dmitrygx** [comment]
bot:pipe:retest

**Comment by dmitrygx** [comment]
bot:pipe:retest

**Comment by dmitrygx** [comment]
@brminich @evgeny-leksikov @hoopoepg could you review pls?

File: src/tools/info/proto_info.c [hunk 1]
```diff
@@ -93,25 +93,75 @@ static void print_resource_usage(const resource_usage_t *usage_before,
...
-                            uint64_t ctx_features,
-                            const ucp_ep_params_t *base_ep_params,
-                            size_t estimated_num_eps, size_t estimated_num_ppn,
-                            unsigned dev_type_bitmap, const char *mem_size)
+static void listener_accept_callback(ucp_ep_h ep, void *arg)
+{
+    ucp_ep_h *server_ep = (ucp_ep_h*)arg;
+
+    *server_ep = ep;
+}
```
Comment (hoopoepg) [inline src/tools/info/proto_info.c:101]: minor: really strange function...
why not just ```*(ucp_ep_h*)arg = ep```

Comment (dmitrygx) [reply src/tools/info/proto_info.c]: fixed

File: src/ucp/core/ucp_ep.h [hunk 2]
```diff
@@ -390,10 +390,11 @@ typedef struct {
  * Endpoint extension for control data path
  */
 typedef struct {
-    ucs_ptr_map_key_t             local_ep_id;   /* Local EP ID */
-    ucs_ptr_map_key_t             remote_ep_id;  /* Remote EP ID */
-    ucp_err_handler_cb_t          err_cb;        /* Error handler */
-    ucp_ep_close_proto_req_t      close_req;     /* Close protocol request */
+    ucp_rsc_index_t          cm_idx; /* CM index */
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.h:393]: unrelated change of moving cm_idx?

File: src/tools/info/proto_info.c [hunk 3]
```diff
@@ -174,38 +222,95 @@ ucs_status_t print_ucp_info(int print_opts,
...
 
     if (print_opts & PRINT_UCP_EP) {
-        status = ucp_worker_get_address(worker, &address, &address_length);
-        if (status != UCS_OK) {
-            printf("<Failed to get UCP worker address>\n");
-            goto out_destroy_worker;
-        }
+        ep_params = *base_ep_params;
+
+        if (strlen(ip_addr) != 0) {
```
Comment (yosefe) [inline src/tools/info/proto_info.c:227]: pls use helper function to create ep

File: src/tools/info/proto_info.c [hunk 4]
```diff
@@ -174,38 +222,95 @@ ucs_status_t print_ucp_info(int print_opts,
...
         if (status != UCS_OK) {
             printf("<Failed to create UCP endpoint>\n");
-            goto out_destroy_worker;
+            goto out_destroy_listener;
+        }
+
+        request_param.op_attr_mask = 0;
+
+        status_ptr = ucp_ep_flush_nbx(ep, &request_param);
+        status     = wait_completion(worker, status_ptr);
```
Comment (yosefe) [inline src/tools/info/proto_info.c:286]: why ep_close is not enough?

File: src/tools/info/ucx_info.c [hunk 5]
```diff
@@ -52,6 +55,7 @@ static void usage() {
     printf("                    'shm'  : shared memory devices only\n");
     printf("                    'net'  : network devices only\n");
     printf("                    'self' : self transport only\n");
+    printf("  -A <ip_address> IP address of the server to connect to\n");
```
Comment (yosefe) [inline src/tools/info/ucx_info.c:58]: `-A <ipv4>   Local network device address to use for creating endpoint in client/server mode`

File: src/tools/info/ucx_info.c [hunk 6]
```diff
@@ -19,6 +19,9 @@
 #include <limits.h>
 
 
+#define MAX_IP_ADDR_LEN 64
```
Comment (yosefe) [inline src/tools/info/ucx_info.c:22]: we should have similar define in sock.h

File: src/tools/info/ucx_info.c [hunk 7]
```diff
@@ -66,6 +70,7 @@ int main(int argc, char **argv)
     size_t ucp_num_ppn;
     unsigned print_opts;
     char *tl_name, *mem_size;
+    char ip_addr[MAX_IP_ADDR_LEN] = "\0";
```
Comment (yosefe) [inline src/tools/info/ucx_info.c:73]: why need to define a string with two null terminators?

File: src/tools/info/ucx_info.h [hunk 8]
```diff
@@ -7,9 +7,12 @@
 #ifndef UCX_INFO_H
 #define UCX_INFO_H
 
+
```
Comment (yosefe) [inline src/tools/info/ucx_info.h:10]: pls remove

File: src/ucp/core/ucp_ep.h (hunk 2 above)
Comment (yosefe) [reply src/ucp/core/ucp_ep.h:393]: why can't we take it from wireup_ep?

File: src/tools/info/proto_info.c [hunk 9]
```diff
@@ -134,32 +134,155 @@ ucs_status_t wait_completion(ucp_worker_h worker, ucs_status_ptr_t status_ptr)
...
+create_listener(ucp_worker_h worker, ucp_listener_h *listener_p,
+                uint16_t *listen_port_p, ucp_ep_h *server_ep_p)
+{
+    ucp_listener_h listener;
+    struct sockaddr_in listen_saddr;
+    ucp_listener_params_t listen_params;
+    ucp_listener_attr_t listen_attr;
+    ucs_status_t status;
+    
+    set_listen_addr(&listen_saddr);
```
Comment (yosefe) [inline src/tools/info/proto_info.c:147]: why need this function?

File: src/tools/info/proto_info.c [hunk 10]
```diff
@@ -134,32 +134,155 @@ ucs_status_t wait_completion(ucp_worker_h worker, ucs_status_ptr_t status_ptr)
     return status;
 }
 
+ucs_status_t
+create_listener(ucp_worker_h worker, ucp_listener_h *listener_p,
+                uint16_t *listen_port_p, ucp_ep_h *server_ep_p)
```
Comment (yosefe) [inline src/tools/info/proto_info.c:139]: ucp_ep_h *server_ep_p -> void *accept_cb_arg

File: src/tools/info/proto_info.c [hunk 11]
```diff
@@ -134,32 +134,155 @@ ucs_status_t wait_completion(ucp_worker_h worker, ucs_status_ptr_t status_ptr)
     return status;
 }
 
+ucs_status_t
```
Comment (yosefe) [inline src/tools/info/proto_info.c:137]: static

File: src/tools/info/proto_info.c [hunk 12]
```diff
@@ -134,32 +134,155 @@ ucs_status_t wait_completion(ucp_worker_h worker, ucs_status_ptr_t status_ptr)
...
+    ucp_address_t *address;
+    size_t address_length;
+    struct sockaddr_in connect_saddr;
+    uint16_t listen_port;
+    ucp_ep_h ep;
+    ucp_request_param_t request_param;
+
+    ep_params = *base_ep_params;
+
+    if (strlen(ip_addr) != 0) {
```
Comment (yosefe) [inline src/tools/info/proto_info.c:204]: maybe pass NULL instead of empty string?

File: src/tools/info/proto_info.c [hunk 13]
```diff
@@ -134,32 +134,155 @@ ucs_status_t wait_completion(ucp_worker_h worker, ucs_status_ptr_t status_ptr)
...
+    ucs_status_ptr_t status_ptr;
+    ucp_ep_params_t ep_params;
+    ucp_address_t *address;
+    size_t address_length;
+    struct sockaddr_in connect_saddr;
+    uint16_t listen_port;
+    ucp_ep_h ep;
+    ucp_request_param_t request_param;
+
+    ep_params = *base_ep_params;
```
Comment (yosefe) [inline src/tools/info/proto_info.c:202]: init inline

File: src/tools/info/proto_info.c [hunk 14]
```diff
@@ -134,32 +134,155 @@ ucs_status_t wait_completion(ucp_worker_h worker, ucs_status_ptr_t status_ptr)
...
+     * configuration is applied */
+    status_ptr = ucp_ep_flush_nbx(ep, &request_param);
+    status     = wait_completion(worker, status_ptr);
+    if (status != UCS_OK) {
+        printf("<Failed to flush UCP endpoint>\n");
+    }
+
+    ucp_ep_print_info(ep, stdout);
+
+    status_ptr = ucp_disconnect_nb(ep);
```
Comment (yosefe) [inline src/tools/info/proto_info.c:248]:
```
/**
 * @ingroup UCP_ENDPOINT
 * @deprecated Replaced by @ref ucp_ep_close_nb.
 */
ucs_status_ptr_t ucp_disconnect_nb(ucp_ep_h ep);
```

File: src/tools/info/proto_info.c [hunk 15]
```diff
@@ -134,32 +134,155 @@ ucs_status_t wait_completion(ucp_worker_h worker, ucs_status_ptr_t status_ptr)
...
+    }
+
+    if (server_ep != NULL) {
+        status_ptr = ucp_ep_close_nb(server_ep, UCP_EP_CLOSE_MODE_FORCE);
+        tmp_status = wait_completion(worker, status_ptr);
+        if ((tmp_status != UCS_OK) && (status == UCS_OK)) {
+            printf("<Failed to destroy server UCP endpoint>\n");
+            status = tmp_status;
+        }
+    }
```
Comment (yosefe) [inline src/tools/info/proto_info.c:262]: pls have a common helper func to blocking-close a ucp endpoint

File: src/tools/info/proto_info.c (hunk 4 above)
Comment (yosefe) [reply src/tools/info/proto_info.c]: ok

File: src/tools/info/ucx_info.c [hunk 16]
```diff
@@ -55,13 +52,15 @@ static void usage() {
     printf("                    'shm'  : shared memory devices only\n");
     printf("                    'net'  : network devices only\n");
     printf("                    'self' : self transport only\n");
-    printf("  -A <ip_address> IP address of the server to connect to\n");
+    printf("  -A <ipv4> Local network device address to use for creating\n"
```
Comment (yosefe) [inline src/tools/info/ucx_info.c:55]: pls align help message indentation

File: src/tools/info/ucx_info.c (hunk 7 above)
Comment (yosefe) [reply src/tools/info/ucx_info.c]: const char *ip_addr = NULL;
...
ip_addr = optarg;

File: src/ucp/core/ucp_ep.h (hunk 2 above)
Comment (dmitrygx) [reply src/ucp/core/ucp_ep.h:393]: > why can't we take it from wireup_ep?

`wireup_ep` doesn't exist at the time when a connection establishment completed

File: src/tools/info/ucx_info.c (hunk 7 above)
Comment (dmitrygx) [reply src/tools/info/ucx_info.c]:
```
const char *ip_addr = NULL;
...
ip_addr = optarg;
```
not sure that this will work if this parameter will go as not the last one

File: src/ucp/core/ucp_ep.h (hunk 2 above)
Comment (yosefe) [reply src/ucp/core/ucp_ep.h:393]: so what do we have? only pure cm lane?

File: src/tools/info/ucx_info.c (hunk 7 above)
Comment (yosefe) [reply src/tools/info/ucx_info.c]: it points int global array of argv

Comment (dmitrygx) [reply src/tools/info/ucx_info.c]: i see, will fix

File: src/ucp/core/ucp_ep.h (hunk 2 above)
Comment (dmitrygx) [reply src/ucp/core/ucp_ep.h:393]: > so what do we have? only pure cm lane?

yes

File: src/tools/info/proto_info.c (hunk 14 above)
Comment (dmitrygx) [reply src/tools/info/proto_info.c]: this code used `ucp_disconnect_nb()` before my changes.
for `server_ep` - I'm using `ucp_ep_close_nb()`
done

File: src/tools/info/proto_info.c (hunk 12 above)
Comment (dmitrygx) [reply src/tools/info/proto_info.c]: done

File: src/tools/info/proto_info.c (hunk 13 above)
Comment (dmitrygx) [reply src/tools/info/proto_info.c]: done

File: src/tools/info/proto_info.c (hunk 9 above)
Comment (dmitrygx) [reply src/tools/info/proto_info.c]: removed

File: src/tools/info/proto_info.c (hunk 10 above)
Comment (dmitrygx) [reply src/tools/info/proto_info.c]: done

File: src/tools/info/proto_info.c (hunk 11 above)
Comment (dmitrygx) [reply src/tools/info/proto_info.c]: done

File: src/tools/info/proto_info.c (hunk 15 above)
Comment (dmitrygx) [reply src/tools/info/proto_info.c]: done

File: src/tools/info/ucx_info.c (hunk 16 above)
Comment (dmitrygx) [reply src/tools/info/ucx_info.c]: done

**Comment by dmitrygx** [comment]
@yosefe @brminich could you review pls?

File: src/tools/info/proto_info.c [hunk 17]
```diff
@@ -134,24 +128,51 @@ ucs_status_t wait_completion(ucp_worker_h worker, ucs_status_ptr_t status_ptr)
...
+{
+    ucp_request_param_t request_param;
+    ucs_status_ptr_t status_ptr;
+    ucs_status_t status;
+
+    if (ep == NULL) {
+        return UCS_OK;
+    }
+
+    if (force) {
```
Comment (yosefe) [inline src/tools/info/proto_info.c:142]: maybe pass flags to this func instead of force=0/1?

File: src/tools/info/proto_info.c [hunk 18]
```diff
@@ -134,24 +128,51 @@ ucs_status_t wait_completion(ucp_worker_h worker, ucs_status_ptr_t status_ptr)
...
 {
     ucp_listener_h listener;
     struct sockaddr_in listen_saddr;
     ucp_listener_params_t listen_params;
     ucp_listener_attr_t listen_attr;
     ucs_status_t status;
-    
-    set_listen_addr(&listen_saddr);
+
+    set_saddr(NULL, 0, &listen_saddr);
```
Comment (yosefe) [inline src/tools/info/proto_info.c:168]: set_saddr("0.0.0.0", 0, &listen_saddr);

File: src/tools/info/proto_info.c [hunk 19]
```diff
@@ -187,47 +208,46 @@ ucs_status_t
...
         ep_params.field_mask |= UCP_EP_PARAM_FIELD_REMOTE_ADDRESS;
-        ep_params.address     = address;
+        ep_params.address     = worker_addr;
     }
 
     status = ucp_ep_create(worker, &ep_params, &ep);
-    if (strlen(ip_addr) == 0) {
-        ucp_worker_release_address(worker, address);
+    if (worker_addr == NULL) {
+        ucp_worker_release_address(worker, worker_addr);
```
Comment (yosefe) [inline src/tools/info/proto_info.c:250]: wierd

File: src/tools/info/proto_info.c [hunk 20]
```diff
@@ -245,20 +265,16 @@ print_ucp_ep_info(ucp_worker_h worker, const ucp_ep_params_t *base_ep_params,
 
     ucp_ep_print_info(ep, stdout);
 
-    status_ptr = ucp_disconnect_nb(ep);
-    tmp_status = wait_completion(worker, status_ptr);
+    tmp_status = ep_close(worker, ep, 0,
+                          (ip_addr != NULL) ? "client" :
+                                              "connected to UCP worker");
```
Comment (yosefe) [inline src/tools/info/proto_info.c:270]: can we set this string as part of the "if" in line 223

File: src/tools/info/proto_info.c [hunk 21]
```diff
@@ -245,20 +265,16 @@ print_ucp_ep_info(ucp_worker_h worker, const ucp_ep_params_t *base_ep_params,
...
-    if (server_ep != NULL) {
-        status_ptr = ucp_ep_close_nb(server_ep, UCP_EP_CLOSE_MODE_FORCE);
-        tmp_status = wait_completion(worker, status_ptr);
-        if ((tmp_status != UCS_OK) && (status == UCS_OK)) {
-            printf("<Failed to destroy server UCP endpoint>\n");
-            status = tmp_status;
-        }
+    tmp_status = ep_close(worker, server_ep, 1, "server");
+    if ((tmp_status != UCS_OK) && (status == UCS_OK)) {
+        status = tmp_status;
```
Comment (yosefe) [inline src/tools/info/proto_info.c:277]: better check ep!=null outside ep_close

File: src/tools/info/proto_info.c (hunk 20 above)
Comment (dmitrygx) [reply src/tools/info/proto_info.c:270]: done

File: src/tools/info/proto_info.c (hunk 21 above)
Comment (dmitrygx) [reply src/tools/info/proto_info.c:277]: done

**Review by yosefe — APPROVED** [review]

File: src/tools/info/proto_info.c [hunk 22]
```diff
@@ -93,25 +93,211 @@ static void print_resource_usage(const resource_usage_t *usage_before,
...
+    ucs_status_ptr_t status_ptr;
+    ucs_status_t status;
+
+    request_param.op_attr_mask = UCP_OP_ATTR_FIELD_FLAGS;
+    request_param.flags        = flags;
+
+    status_ptr = ucp_ep_close_nbx(ep, &request_param);
+    status     = wait_completion(worker, status_ptr);
+    if (status != UCS_OK) {
+        printf("<Failed to destroy %s UCP endpoint>\n", ep_type);
```
Comment (evgeny-leksikov) [inline src/tools/info/proto_info.c:144]: destroy must be completed for any status excepting "in progress", return code of ep_close indicates status ro the peer, I think better to ignore it.

File: src/tools/info/proto_info.c [hunk 23]
```diff
@@ -93,25 +93,211 @@ static void print_resource_usage(const resource_usage_t *usage_before,
...
+    if (status != UCS_OK) {
+        printf("<Failed to flush UCP endpoint>\n");
+    }
+
+    ucp_ep_print_info(ep, stdout);
+
+    tmp_status = ep_close(worker, ep, 0, ep_name);
+    if ((tmp_status != UCS_OK) && (status == UCS_OK)) {
+        status = tmp_status;
+    }
```
Comment (evgeny-leksikov) [inline src/tools/info/proto_info.c:265]: IMO not needed

File: src/tools/info/proto_info.c [hunk 24]
```diff
@@ -93,25 +93,211 @@ static void print_resource_usage(const resource_usage_t *usage_before,
...
+
+    if (server_ep != NULL) {
```
Comment (evgeny-leksikov) [inline src/tools/info/proto_info.c:267]: I think we can check ip_addr for consistency with code above and just assert server_ep

File: src/tools/info/proto_info.c [hunk 25]
```diff
@@ -93,25 +93,211 @@ static void print_resource_usage(const resource_usage_t *usage_before,
...
+        tmp_status = ep_close(worker, server_ep, UCP_EP_CLOSE_FLAG_FORCE,
+                              "server");
+        if ((tmp_status != UCS_OK) && (status == UCS_OK)) {
+            status = tmp_status;
+        }
```
Comment (evgeny-leksikov) [inline src/tools/info/proto_info.c:272]: the same, not needed

File: src/tools/info/proto_info.c [hunk 26]
```diff
@@ -93,25 +93,211 @@ static void print_resource_usage(const resource_usage_t *usage_before,
...
+        if (status != UCS_OK) {
+            return status;
+        }
+
+        ucs_strncpy_zero(ep_name, "client", sizeof(ep_name));
+
+        set_saddr(ip_addr, listen_port, &connect_saddr);
+
+        ep_params.field_mask      |= UCP_EP_PARAM_FIELD_FLAGS |
+                                     UCP_EP_PARAM_FIELD_SOCK_ADDR;
```
Comment (evgeny-leksikov) [inline src/tools/info/proto_info.c:227]: maybe add another option to enable error handling? sock_addr usually is used together with err handling mode peer

File: src/tools/info/ucx_info.c [hunk 27]
```diff
@@ -52,12 +52,15 @@ static void usage() {
     printf("                    'shm'  : shared memory devices only\n");
     printf("                    'net'  : network devices only\n");
     printf("                    'self' : self transport only\n");
+    printf("  -A <ipv4>       Local IPv4 device address to use for creating\n"
+           "                  endpoint in client/server mode");
```
Comment (evgeny-leksikov) [inline src/tools/info/ucx_info.c:56]: maybe add IPv6 support as well?

File: src/tools/info/proto_info.c (hunk 26 above)
Comment (dmitrygx) [reply src/tools/info/proto_info.c:227]: we already have `e` in features, it is set as part of `base_ep_params`
```
...
ucp_ep_params_t ep_params  = *base_ep_params;
...
```

File: src/tools/info/proto_info.c (hunk 22 above)
Comment (dmitrygx) [reply src/tools/info/proto_info.c:144]: peer is on the same process, so I'd expect that it shouldn't fail

Comment (evgeny-leksikov) [reply src/tools/info/proto_info.c]: anyway better to follow the doc, IMO this error message is wrong

Comment (dmitrygx) [reply src/tools/info/proto_info.c]: https://github.com/openucx/ucx/blob/b7ddc45fccd955b604f28abd948c3f8b7e03c4c7/examples/ucp_client_server.c#L471
the code is almost the same that we have in `ucp_client_server` example

Comment (evgeny-leksikov) [reply src/tools/info/proto_info.c]: then I think we have to fix there too :)

File: src/tools/info/proto_info.c (hunk 24 above)
Comment (evgeny-leksikov) [reply src/tools/info/proto_info.c:268]: > (e.g. wrong IP:port specified), so flush completed with error

this case should be handled earlier

> Added ucs_assert(ip_addr != NULL) inside server_ep != NULL

ok

File: src/tools/info/proto_info.c (hunk 22 above)
Comment (dmitrygx) [reply src/tools/info/proto_info.c]: fixed

File: src/tools/info/proto_info.c [hunk 28]
```diff
@@ -93,25 +93,200 @@ static void print_resource_usage(const resource_usage_t *usage_before,
...
+    request_param.op_attr_mask = 0;
+    /* do EP flush to make sure that fully completed to a peer and final
+     * configuration is applied */
+    status_ptr = ucp_ep_flush_nbx(ep, &request_param);
+    status     = wait_completion(worker, status_ptr);
+    if (status != UCS_OK) {
+        printf("<Failed to flush UCP endpoint>\n");
+    }
+
+    ucp_ep_print_info(ep, stdout);
```
Comment (evgeny-leksikov) [inline src/tools/info/proto_info.c:254]: does it make sense to print failed EP?

Comment (dmitrygx) [reply src/tools/info/proto_info.c:256]: fixed

**Comment by dmitrygx** [comment]
@evgeny-leksikov could you review pls?

**Review by evgeny-leksikov — APPROVED** [review]

**Review by yosefe — APPROVED** [review]

**Comment by dmitrygx** [comment]
@yosefe could you review the 2nd commit pls (it contains changes after @evgeny-leksikov review)?

**Comment by dmitrygx** [comment]
@yosefe thanks!
should I squash?

**Comment by yosefe** [comment]
> @yosefe thanks!
> should I squash?

LGTM both commits, pls squash

**Comment by dmitrygx** [comment]
bot:pipe:retest

**Review by yosefe — APPROVED** [review]
